### PR TITLE
bulkio: Implement `SHOW SCHEDULES`

### PR DESCRIPTION
**Pull request by miretskiy**
Informs #51850
Informs #51600

Implement `SHOW SCHEDULES` statemen which displays the information
on scheduled jobs.

Display schedule information, optionally filtered
by schedule state (paused or not) and optionally restricted
just to the backup schedules:

```
SHOW [RUNNING|PAUSED] SCHEDULES [FOR BACKUP]
```

In addition, it is possible to display information
for a specific schedule:

```
SHOW SCHEDULE 123
```

Release Notes (enterprise change): `SHOW SCHEDULES` displays
information about the scheduled jobs.

### CODE REVIEW

File: pkg/sql/sem/tree/show.go [hunk 1]
```diff
@@ -533,3 +533,92 @@ func (node *ShowPartitions) Format(ctx *FmtCtx) {
...
+	// only show those schedules that are currently active.
+	ActiveSchedules
+
+	// PausedSchedules indicates that show schedules should
+	// only show those schedules that are currently paused.
+	PausedSchedules
+)
+
+// String implements Stringer interface.
+func (s ScheduleState) String() string {
```
Comment (rohany) [inline pkg/sql/sem/tree/show.go:587]: [nit]: we usually implement `Format` for these things rather than string()

File: pkg/sql/sem/tree/show.go [hunk 2]
```diff
@@ -533,3 +533,92 @@ func (node *ShowPartitions) Format(ctx *FmtCtx) {
...
+	// the execution of the scheduled backups.
+	ScheduledBackupExecutor
+)
+
+// InternalName returns an internal executor name.
+// This name can be used to filter matching schedules.
+func (t ScheduledJobExecutorType) InternalName() string {
+	switch t {
+	case ScheduledBackupExecutor:
+		return "scheduled-backup-executor"
```
Comment (rohany) [inline pkg/sql/sem/tree/show.go:555]: can you make these constants

File: pkg/sql/parser/sql.y [hunk 3]
```diff
@@ -4226,6 +4236,58 @@ show_jobs_stmt:
   }
 | SHOW JOB error // SHOW HELP: SHOW JOBS
 
+// %Help: SHOW SCHEDULES - list periodic schedules
```
Comment (rohany) [inline pkg/sql/parser/sql.y:4239]: add a test in `TextContextualHelp`

File: pkg/sql/delegate/show_schedules.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
+)
+
+// commandColumn converts executor execution arguments into jsonb representation.
+const commandColumn = `crdb_internal.pb_to_json('cockroach.jobs.jobspb.ExecutionArguments', execution_args)->'args'`
```
Comment (rohany) [inline pkg/sql/delegate/show_schedules.go:23]: hah, i see that your work on pb_to_json was not all done out of generosity :)

**Review by adityamaru — COMMENTED** [review]
> Reviewed 11 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52230)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[pkg/sql/parser/sql.y, line 4244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDoP4G11j4QghkSt7zh:-MDoP4G11j4QghkSt7zi:bbfi6ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/parser/sql.y#L4244)):*
> ```
> SCHEDUES
> ```
nit: SCHEDULES
___
*[pkg/sql/sqlbase/constants.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDoReLd6bZ3dc-HJork:-MDoReLd6bZ3dc-HJorl:b-xrdh5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/sqlbase/constants.go#L91)):*
> ```
> CrdbInternalSchedulesTableID
> ```
Out of curiosity, where is this used? I can't seem to find it.

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52230)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @rohany)

___
*[pkg/sql/delegate/show_schedules.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDk8AS3s0G7XtLORnYp-r1-23:-MDoUONgE7_Ka2KslHwa:b-atnf7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/delegate/show_schedules.go#L23)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

hah, i see that your work on pb_to_json was not all done out of generosity :)
</blockquote></details>

Yes; I really did not want to use crdb_internal table approach used by show jobs.  I found
it pretty bad that we scan entire table, decode every proto message, just to then filter out some things we actually want to show (predicate pushdown notwithstanding).

I want to use optimizer! And that's why I started down the internal function route, which evolved into pb_to_json.
___
*[pkg/sql/parser/sql.y, line 4239 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDk8AS3s0G7XtLORnYk-r1-4239:-MDoUxu735gUdv87QiFf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/parser/sql.y#L4239)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

add a test in `TextContextualHelp`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/show.go, line 555 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDk8AS3s0G7XtLORnYl-r1-555:-MDoct8uB1nvgWOrQ8jn:b-a6llfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/sem/tree/show.go#L555)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

can you make these constants
</blockquote></details>

Made a map lookup name instead.
___
*[pkg/sql/sem/tree/show.go, line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDk8AS3s0G7XtLORnYl-r1-587:-MDodd7L7K-g03PF2P1g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/sem/tree/show.go#L587)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: we usually implement `Format` for these things rather than string()
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/constants.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDoReLd6bZ3dc-HJork:-MDoditP9zrX33Yb1ZN9:b-ccist5) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/sqlbase/constants.go#L91)):*
> ```
> sqltelemetry.IncrementShowCounter(sqltelemetry.Schedules)
> ```
Uses a map w/ this constant.

File: pkg/sql/parser/parse_test.go [hunk 5]
```diff
@@ -2028,16 +2037,10 @@ $function$`,
 
 		{`CANCEL JOB a`, `CANCEL JOBS VALUES (a)`},
 		{`EXPLAIN CANCEL JOB a`, `EXPLAIN CANCEL JOBS VALUES (a)`},
-		{`CANCEL JOBS FOR SCHEDULE a`, `CANCEL JOBS FOR SCHEDULES VALUES (a)`},
```
Comment (rohany) [inline pkg/sql/parser/parse_test.go:2031]: Why did you remove these?

**Review by rohany — APPROVED** [review]

File: pkg/sql/parser/parse_test.go (hunk 5 above)
Comment (miretskiy) [reply pkg/sql/parser/parse_test.go:2031]: A mistake.  Thanks.

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52230)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @rohany)

___
*[pkg/sql/sqlbase/constants.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52230#-MDoReLd6bZ3dc-HJork:-MDpN7Mn8hVjfMs5Tjn_:bnqvy57) ([raw file](https://github.com/cockroachdb/cockroach/blob/89764dfc893d60d8feb475a66c2bad06d265981c/pkg/sql/sqlbase/constants.go#L91)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> sqltelemetry.IncrementShowCounter(sqltelemetry.Schedules)
> ```
Uses a map w/ this constant.
</blockquote></details>

Ooops -- turns out, this constant was a left over (and was causing logic test to fail).  Reverted.

**Comment by rohany** [comment]
lgtm
